### PR TITLE
fqdn: Update high-level package docs

### PR DESCRIPTION
**Pull request by raybejjani**
I'm not too sure if all of this information should go in the package docs. We don't have a place for developer oriented debugging notes, for example, but a lot of what's in here is also stated in the cilium docs around these policy components.

Edit: reviewers, I added the complementary debug doc as well. You can skip it if you'd like but it's there so you have a sense for it when considering what information is where.

### CODE REVIEW

**Comment by raybejjani** [comment]
test-docs-please

**Comment by raybejjani** [comment]
test-docs-please

**Comment by raybejjani** [comment]
test-docs-please

File: Documentation/contributing/development/debugging.rst [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+returned are not compatible with our datapath implementation. Sometimes we also
+just have bugs.
+
+
+Isolating the source of toFQDNs issues
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While there is no common culprit when debugging, the DNS Proxy is the most
+one-off component in this chain. The cascading caching scheme is also complex
+in its behaviour. Isolating an issue to these components, or to the policy
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: > Isolating an issue to these components

to

> Isolating an issue **in** these components

File: Documentation/contributing/development/debugging.rst [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+layer or datapath is often the first step when debugging toFQDNs related
+issues. Generally, working top-down is easiest as the information needed to
+verify low-level correctness can be collected in the initial debug invocations.
+
+*REFUSED vs  NXDOMAIN responses*
+The proxy uses REFUSED DNS responses to indicate a denied request. Some libc
+implementations, notably musl which is common in alpine images, terminate the
+whole DNS search in these cases. To work around it, denied responses can be
+configured to be NXDOMAIN via ``--tofqdns-dns-reject-response-code``.
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: change

> via

to

> by setting the command line argument

File: Documentation/contributing/development/debugging.rst [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+*Monitor Events*
+The DNS Proxy emits L7 monitor events. One for the request and one for the
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: Assuming I understand, I would reword to this:

The DNS Proxy emits multiple L7 monitor events. One for the request and one for the response (if allowed). Often the L7 DNS rules are paired with L3 toFQDNs rules and, thus, emits extra events.

File: Documentation/contributing/development/debugging.rst [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+response, if allowed. Often the L7 DNS rules are paired with L3 toFQDNs rules
+and events are emitted. 
+
+
+.. Note::
+
+    Bu sure to run cilium monitor on the same node as the pod being debugged!
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: Be

File: Documentation/contributing/development/debugging.rst [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  misbehaves and many pods see DNS timeouts. This isn't an actual problem with
+  cilium or the proxy although it can be caused by policy blocking the DNS
+  target server if it is in-cluster.
+- ``Timed out waiting for datapath updates of FQDN IP information; returning
+  response``: When the proxy updates the DNS caches with response data, it
+  needs to allow some time for that information to get into the datapath.
+  Otherwise, pods would attempt to make the outbound connection (the thing that
+  caused the DNS lookup) before the datapath is ready. Many stacks retry the
+  SYN in such cases but some return an error and some apps further crash as a
+  response. This delay is configurable via
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: by setting the command argument

File: Documentation/contributing/development/debugging.rst [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+entry with Security Identity ``16777217``. The L7 rule, used to redirect to the
+DNS Proxy is also present with a populated ``PROXY PORT``. It has a 0
+``IDENTITY`` as it is an L3 wildcard.
+
+An identity missing here can be an error in various places:
+- Policy doesn't actually allow this Endpoint to connect. A sanity check is to
+use ``cilium endpoint list`` to see if cilium thinks it should have policy
+enforcement.
+- Endpoint regeneration is slow and the Policy Map has not been updated yet.
+This can occur in cases where we have leaked IPs from the DNS cache (i.e. the
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: they were never

File: Documentation/contributing/development/debugging.rst [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+were never deleted correctly) or when there are legitimately many IPs. It can
+also simply mean an overloaded node or even a deadlock within cilium.
```
Comment (nathanjsweet) [inline Documentation/contributing/development/debugging.rst]: In this case, is the deadlock in cilium you are suggesting the result of resource constraints?

File: pkg/fqdn/doc.go [hunk 8]
```diff
@@ -12,20 +12,238 @@
...
+// (https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-config)
+// In the example above `matchPattern: "*"` allows all requests and makes
+// `matchName: "cilium.io"` redundant.
+// Notes:
+//  - The forwarded requests are sent from cilium-agent on the host interface
+//    and not from the endpoint
+//  - Users must explicitly allow `*.*.svc.cluster.local.` in k8s clusters.
+//    This is not automatic
+//  - L7 DNS rules are egress-only
+//  - The proxy emits L7 cilium-monitor events: one for the request and a
```
Comment (nathanjsweet) [inline pkg/fqdn/doc.go]: one for the request, an accept/reject event, and the final response.

File: pkg/fqdn/doc.go [hunk 9]
```diff
@@ -12,20 +12,238 @@
...
+//    accept/reject event, and the final response
+//
+// Apart from allowing or denying DNS requests the DNS proxy is used to observe
```
Comment (nathanjsweet) [inline pkg/fqdn/doc.go]: Apart from allowing or denying DNS requests*,* the DNS proxy is used to observe

File: pkg/fqdn/doc.go [hunk 10]
```diff
@@ -12,20 +12,238 @@
...
+// way and can overlap with toCIDR rules in policiies. They are placed in the
+// node-local ipcache and in the policy map of each endpoint that is allowed to
+// connect to them (i.e. defined in the L3 DNS rule).
+// Notes:
+// - Generally speaking, toFQDNs can only handle non-cluster IPs. In-cluster
+//   policy should use toEndpoints and toServices. This is partly historical but
+//   is because of ipcache limitations when mapping ip->identity. Endpoint
+//   identities can clobber the FQDN IP identity
+// - Despite being tracked per-Endpoint. DNS lookup IPs are collected into a
+//   global cache. This is historical and can be changed
```
Comment (nathanjsweet) [inline pkg/fqdn/doc.go]: missing period.

File: pkg/fqdn/doc.go [hunk 11]
```diff
@@ -12,20 +12,238 @@
...
+// DNS requests are distinct traffic from the connections that pods make with
+// the response information. This makes it difficult to correlate one DNS
+// lookup to a later connection; a pod may reuse the IPs in a DNS response an
+// arbitrary time after the lookup occurred, even past the DNS TTL. The
+// solution is multi-layered for historical reasons:
+// - Keep a per-Endpoint cache that can be stored to disk and restored on
+//   startup. These caches apply TTL expiration and limit the IP count per domain.
+// - Keep a global cache to combine all this DNS information and send it to the
+//   policy system. This cache applies TTL but not per-domain limits.
+//   This causes a DNS lookup in one endpoint to leak to another!
```
Comment (nathanjsweet) [inline pkg/fqdn/doc.go:141]: :thinking:

File: pkg/fqdn/doc.go [hunk 12]
```diff
@@ -12,20 +12,238 @@
...
+//   churn. This is mostly history as CIDR identities now have a near-0
+//   allocation overhead
+// - DNSCache deletes only currently occur when the cilium API clears the cache
+//   or when the garbage collector evicts entries
+// - The combination of caches: per-Endpoint and global must manage disparate
+//   behaviours of pods. The worst case scenario is one where one pod makes many
+//   requests to a target with changing IPs (like S3) but another makes few
+//   requests that are long-lived. We need to ensure "fairness" where one does
+//   not starve the other. The limits in the per-Endpoint caches allow this, and
+//   the global cache acts are a collector across different Endpoints (without
```
Comment (nathanjsweet) [inline pkg/fqdn/doc.go]: acts as?

**Review by nathanjsweet — CHANGES_REQUESTED** [review]
I hope my comments are helpful. Very informative!

File: Documentation/contributing/development/debugging.rst (hunk 7 above)
Comment (raybejjani) [reply Documentation/contributing/development/debugging.rst]: no! an actual lock vs lock deadlock. We have had those because we have a bunch of concurrent stuff that synchronises with a combination of locks, channels and wait groups so it can get messy.

File: pkg/fqdn/doc.go (hunk 10 above)
Comment (raybejjani) [reply pkg/fqdn/doc.go]: oh, this was an experiment on my part. I usually have proper periods in lists but I think I read somewhere that one isn't supposed to... so I tried it out. I actually did try to look it up and I think I'll add them back in to satisfy my OCD.

File: pkg/fqdn/doc.go (hunk 11 above)
Comment (raybejjani) [reply pkg/fqdn/doc.go:141]: There were many hand-wringey converstions about this. I think the conclusion was that:
1- Trusting the DNS data itself would use other tools (including controlling the DNS lookups and where they can go)
2- It's a little messy anyway because the DNS TTL is much shorter than a connection could be and we would expect to allow the whole connection. So letting it leak between endpoints is sort of like that (hand-wave).

Now that we have more wiring to the connection tracking we can clamp down on this quite a bit. Setting aside weird case like one pod doing service lookup and giving the IP to another (which I planned to support as a sibling to the DNS discovery) it's unlikely that leaking out of a pod will be needed.

**Review by raybejjani — COMMENTED** [review]
Thanks @nathanjsweet! I basically took all your comments in as-is. I reworded some bits.

**Comment by raybejjani** [comment]
test-docs-please

File: pkg/fqdn/doc.go [hunk 13]
```diff
@@ -12,20 +12,238 @@
...
+// - The combination of caches: per-Endpoint and global must manage disparate
+//   behaviours of pods. The worst case scenario is one where one pod makes many
+//   requests to a target with changing IPs (like S3) but another makes few
+//   requests that are long-lived. We need to ensure "fairness" where one does
+//   not starve the other. The limits in the per-Endpoint caches allow this, and
+//   the global cache acts as a collector across different Endpoints (without
+//   restrictions).
+//
+// Expiration of DNS data is handled by the dns-garbage-collector-job controller.
+// Historically, the only expiraion was TTL based and the per-Endpoint and
```
Comment (jrajahalme) [inline pkg/fqdn/doc.go]: "expiration"

File: Documentation/contributing/development/debugging.rst [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+  SYN in such cases but some return an error and some apps further crash as a
+  response. This delay is configurable by setting the
+  ``--tofqdns-proxy-response-max-delay`` command line argument but defaults to
+  100ms. It can be exceeded if the system is under load.
+
+
+Identities and Policy
+~~~~~~~~~~~~~~~~~~~~~
+
+Once a DNS response has been passed back through the proxy and is places in the
```
Comment (jrajahalme) [inline Documentation/contributing/development/debugging.rst]: "placed"

**Review by jrajahalme — APPROVED** [review]
Nice! Noticed two typos though :-)

**Comment by raybejjani** [comment]
test-docs-please

**Review by nebril — APPROVED** [review]
LGTM

File: Documentation/contributing/development/debugging.rst [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+returned are not compatible with our datapath implementation. Sometimes we also
+just have bugs.
+
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Isolating the source of toFQDNs issues
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While there is no common culprit when debugging, the DNS Proxy is the most
+one-off component in this chain. The cascading caching scheme is also complex
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]: I'm not quite sure what `the most one-off component` means.

File: Documentation/contributing/development/debugging.rst [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+toFQDNs related issues. Generally, working top-down is easiest as the
+information needed to verify low-level correctness can be collected in the
+initial debug invocations.
+
+
+REFUSED vs NXDOMAIN responses
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The proxy uses REFUSED DNS responses to indicate a denied request. Some libc
+implementations, notably musl which is common in alpine images, terminate the
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
implementations, notably musl which is common in Alpine Linux images, terminate the
```

File: Documentation/contributing/development/debugging.rst [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+whole DNS search in these cases. To work around it, denied responses can be
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]: What is the user-facing error in these cases? Could we provide an example of what this looks like?

File: Documentation/contributing/development/debugging.rst [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+The above is for a simple ``curl cilium.io`` in a pod. The L7 DNS request is
+the first set of message and the subsequent L3 connection is the HTTP
+component. AAAA DNS lookups commonly happen but were removed to simplify the
+example.
+
+- If no L7 DNS requests appear, the proxy redirect is not in place. This may
+  mean that the policy does not select this endpoint or there is an issue with
+  the proxy redirection.
+  In the past, a bug occurred with more permissive L3 rules overriding the
+  proxy redirect, causing the proxy to never see the requests.
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst:94]: Note, users can check whether there are redirects using `cilium status --all-redirects`.

File: Documentation/contributing/development/debugging.rst [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+- If the L7 DNS request is blocked, with an explicit denied message, then the
+  requests are not allowed by the proxy. This may be due to a typo or the
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
  requests are not allowed by the proxy. This may be due to a typo in the network policy, or the
```

File: Documentation/contributing/development/debugging.rst [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+  matchPattern rule not being correct. It can also be due to a bug in policy
+  propagation to the DNS Proxy.
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]: Perhaps, "If you believe it is a bug, please <file an issue>"?

File: Documentation/contributing/development/debugging.rst [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+- If the DNS request is allowed, with an explicit message, and it should not
+  be, this may be because a more general policy is in place that allows the
+  request. ``matchPattern: "*"`` visibility policies are commonly in place and
+  would supersede all other, more restrictive, policies.
+  If no other policies are in place, incorrect allows may indicate a bug when
+  passing policy information to the proxy. There is no way to dump the rules in
+  the proxy, but a debug log is printed when a rule is added. Look for 
+  ``DNS Proxy updating matchNames in allowed list during UpdateRules``.
+  See pkg/proxy/dns.go
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]: This last fragment doesn't seem fully formed. Maybe..

```suggestion
  The pkg/proxy/dns.go file contains the DNS proxy implementation.
```

File: Documentation/contributing/development/debugging.rst [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+~~~~~~~~~~~~~~~~~~~~~
+
+Once a DNS response has been passed back through the proxy and is placed in the
+DNS cache ``toFQDNs`` rules can begin using the IPs in the cache. There are
+multiple layers of cache:
+
+- A per-Endpoint ``DNSCache`` stores the lookups for this endpoint. It is
+  restored on cilium startup with the endpoint. Limits are applied here for
+  ``--tofqdns-endpoint-max-ip-per-hostname`` and TTLs are tracked. The
+  ``--tofqdns-min-ttl`` is not used here
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
  ``--tofqdns-min-ttl`` is not used here.
```

File: Documentation/contributing/development/debugging.rst [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+    Ingress     1          ANY          NONE         0       0
+    Egress      0          53/TCP       36447        0       0
+    Egress      0          53/UDP       36447        138     2
+    Egress      16777217   ANY          NONE         477     6
+
+
+L3 ``toFQDNs`` rules are egress only, so we would expect to see an ``Egress``
+entry with Security Identity ``16777217``. The L7 rule, used to redirect to the
+DNS Proxy is also present with a populated ``PROXY PORT``. It has a 0
+``IDENTITY`` as it is an L3 wildcard.
```
Comment (joestringer) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
``IDENTITY`` as it is an L3 wildcard, ie the policy allows any peer on the specified port.
```

File: pkg/fqdn/doc.go [hunk 24]
```diff
@@ -12,20 +12,238 @@
...
+// global caches would expire data at the same time without added logic.
+// This is not true when we apply per-host IP limits in the cache. These
+// default to 50 IPs for a given domain, per Endpoint. To account for these
+// evictions the controller handles TTL and IP limit evictions. This ensures
+// that the global cache is consistent with the per-Endpoint caches. The result
+// is that the actual expiration is imprecise (TTL especially). The caches mark
+// to-evict data internally and only do so on GC method calls from the
+// controller.
+// When DNS data is evicted from any per-Endpoint cache, for any reason, each
+// IP is retained as a "zombie". These are tracked per IP for that endpoint
```
Comment (joestringer) [inline pkg/fqdn/doc.go]: Not sure if we can work a little more clarity into here, something like:

`These "zombies" represent IPs that were previously associated with a resolved DNS name, but the DNS name is no longer known (for example because of TTL expiry). However there may still be an active connection associated with the zombie IP.`

(then the following part describes alive/dead zombies)

File: pkg/fqdn/doc.go [hunk 25]
```diff
@@ -12,20 +12,238 @@
...
+// (with a limit of 10000 by default). When the Connection Tracking garbage
+// collector runs, it marks any zombie IP that correlates to a live connection
+// by that endpoint as "alive". At the next iteration of the
+// dns-garbage-collector-job controller, the not-live zombies are finally
+// evicted. These IPs are then, finally, no longer placed into the global cache
+// on behalf of this endpoint. Other endpoints may have live DNS TTLs or
+// connections to the same IPs, however, so these IPs may be inserted into the
+// global cache for the same domain or a different one (or both).
+// Note: The CT GC has a variable run period. This ranges from 30s to 12 hours
+// and is shorter when more connection churn is observed.
```
Comment (joestringer) [inline pkg/fqdn/doc.go]: minor nit for maintainability, I wonder for these constants like 1000, 30s, 12h whether we should refer to the variable that holds these configurations?

**Review by joestringer — APPROVED** [review]
Awesome docs.

I wonder on a couple of the points how well the docs will date (particularly things that reference history without a reference point to say when this information was true), but I don't really have any good suggestions how we can improve this. AFAIK all of this will stay true for the forseable future, as we're not actively planning any work at the moment on the core DNS handling engine.

Feel free to either address the comments below in this PR or follow up with the minor fixups.

File: Documentation/contributing/development/debugging.rst [hunk 26]
```diff
@@ -0,0 +1,320 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    http://docs.cilium.io
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
    https://docs.cilium.io
```
We switched to the `https` version for this URL in all the documentation a few days ago

File: Documentation/contributing/development/debugging.rst [hunk 27]
```diff
@@ -0,0 +1,320 @@
...
+#########
+
+
+toFQDNs and DNS Debugging
+-------------------------
+
+The interactions of L3 toFQDNs and L7 DNS rules can be difficult to debug
+around. Unlike many other policy rules, these are resolved at runtime with
+unknown data. Pods may create large numbers of IPs in the cache or the IPs
+returned are not compatible with our datapath implementation. Sometimes we also
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
returned may not be compatible with our datapath implementation. Sometimes we also
```

File: Documentation/contributing/development/debugging.rst [hunk 28]
```diff
@@ -0,0 +1,320 @@
...
+  per-Endpoint cache but are waiting for the Connection Tracking GC to mark
+  them in-use or not. This can take up to 12 hours to occur. This list is
+  size-limited by ``--tofqdns-max-deferred-connection-deletes``. 
+- A global ``DNSCache`` where all endpoint and poller DNS data is collected. It
+  does apply the ``--tofqdns-min-ttl`` value but not the
+  ``--tofqdns-endpoint-max-ip-per-hostname`` value.
+
+If an IP exists in the FQDN cache (check with ``cilium fqdn cache list``) then
+``toFQDNs`` rules that select a domain name, either explicitly via
+``matchName`` or via ``matchPattern`` should cause IPs for that domain to have
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
``matchName`` or via ``matchPattern``, should cause IPs for that domain to have
```

File: Documentation/contributing/development/debugging.rst [hunk 29]
```diff
@@ -0,0 +1,320 @@
...
+    323        k8s:class=xwing
+               k8s:io.cilium.k8s.policy.cluster=default
+               k8s:io.cilium.k8s.policy.serviceaccount=default
+               k8s:io.kubernetes.pod.namespace=default
+               k8s:org=alliance
+    ...
+    16777217   cidr:104.198.14.52/32
+               reserved:world
+
+Note that CIDR identities are allocated locally on the node and have a high-bit set so they are often in the 16 Million range.
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]:
```suggestion
Note that CIDR identities are allocated locally on the node and have a high-bit set so they are often in the 16-million range.
```

Also, break the long line for consistency?

File: Documentation/contributing/development/debugging.rst [hunk 30]
```diff
@@ -0,0 +1,320 @@
...
+          - matchPattern: "*"
+
+The L7 DNS rule has an implicit L3 allow-all because it defines only L4 and L7
+sections. This is the second selector in the list, and includes all possible L3
+identities known in the system. In contrast, the first selector, which
+corresponds to the ``toFQDNS: matchName: "*"`` rule would list all identities
+for IPs that came from the DNS Proxy. Other CIDR identities would not be
+included.
+
+*Datapath Plumbing*
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]: Does Sphinx render this as a paragraph title? I'd expect it to joint it with the next line.

File: Documentation/contributing/development/debugging.rst [hunk 31]
```diff
@@ -0,0 +1,320 @@
...
+                                                                                                                             15194
+                                                                                                                             18892
+                                                                                                                             25379
+                                                                                                                             29200
+                                                                                                                             32255
+                                                                                                                             33831
+                                                                                                                             16777217
+    &LabelSelector{MatchLabels:map[string]string{reserved.none: ,},MatchExpressions:[]LabelSelectorRequirement{},}   1
+
+In this example 16777217 is used by two selectors, one with ``matchPattern: "*"`` and another empty one. This is because of the policy in use:
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]: Break the line?

File: Documentation/contributing/development/debugging.rst [hunk 32]
```diff
@@ -0,0 +1,320 @@
...
+    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium bpf policy get 3459
+    DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS
+    Ingress     reserved:unknown              ANY          NONE         1367    7
+    Ingress     reserved:host                 ANY          NONE         0       0
+    Egress      reserved:unknown              53/TCP       36447        0       0
+    Egress      reserved:unknown              53/UDP       36447        138     2
+    Egress      cidr:104.198.14.52/32         ANY          NONE         477     6
+                reserved:world 
+
+Note that the labels for identities are resolved here. This can be skipped, or there may be cases where this doesn't occur:
```
Comment (qmonnet) [inline Documentation/contributing/development/debugging.rst]: Break the line?

**Review by nathanjsweet — APPROVED** [review]

File: Documentation/contributing/development/debugging.rst (hunk 15 above)
Comment (raybejjani) [reply Documentation/contributing/development/debugging.rst]: I had "most likely to have bugs" in mind but I changed the wording to say that.

File: Documentation/contributing/development/debugging.rst (hunk 17 above)
Comment (raybejjani) [reply Documentation/contributing/development/debugging.rst]: It's app dependent, I'm afraid. The return from the lookup function is empty and the app would then error. I added a bit more explanation of what to look for.

File: Documentation/contributing/development/debugging.rst (hunk 20 above)
Comment (raybejjani) [reply Documentation/contributing/development/debugging.rst]: This was more to allude to the section related to this. There's some basic problem isolation around the policy part later.

File: Documentation/contributing/development/debugging.rst (hunk 30 above)
Comment (raybejjani) [reply Documentation/contributing/development/debugging.rst]: Nope! I had placeholders for the sections and I forgot to switch this one. Good catch!

File: pkg/fqdn/doc.go (hunk 25 above)
Comment (raybejjani) [reply pkg/fqdn/doc.go]: I considered it but I erred on being lazy. These would be the defaults like `defaults.ToFQDNsMaxDeferredConnectionDeletes`. I wasn't sure how valuable that would be since the names are a tad opaque and finding them would be straightforward. The actual value seemed more important to me, since knowing that something may take 12 hours would change the debugging choices. I added the references in with the numbers.

**Comment by raybejjani** [comment]
test-docs-please

File: pkg/fqdn/doc.go (hunk 25 above)
Comment (joestringer) [reply pkg/fqdn/doc.go]: I was thinking about someone editing the default value, then wanting to grep to find references, so having the variable there should help keeping these docs up to date in such a case (hopefully).
